core/src/main/java/io/undertow/server/handlers/PathTemplateHandler.java

```diff
@@ -20,85 +20,155 @@
+        Objects.requireNonNull(next);
+
         this.rewriteQueryParameters = rewriteQueryParameters;
-        this.next = next;
+        builder = PathTemplateRouter.SimpleBuilder.newBuilder(next);
+        router = builder.build();
     }
 
-    @Override
-    public void handleRequest(HttpServerExchange exchange) throws Exception {
-        PathTemplateMatcher.PathMatchResult<HttpHandler> match = pathTemplateMatcher.match(exchange.getRelativePath());
-        if (match == null) {
-            next.handleRequest(exchange);
-            return;
-        }
-        exchange.putAttachment(PATH_TEMPLATE_MATCH, new PathTemplateMatch(match.getMatchedTemplate(), match.getParameters()));
-        exchange.putAttachment(io.undertow.util.PathTemplateMatch.ATTACHMENT_KEY, new io.undertow.util.PathTemplateMatch(match.getMatchedTemplate(), match.getParameters()));
-        if (rewriteQueryParameters) {
-            for (Map.Entry<String, String> entry : match.getParameters().entrySet()) {
-                exchange.addQueryParam(entry.getKey(), entry.getValue());
-            }
+    /**
+     * Adds a template and handler to the underlying router.
+     *
+     * @param uriTemplate The URI path template.
+     * @param handler     The handler to use for requests that match the specified template.
+     *
+     * @return Reference to this handler.
+     */
+    public PathTemplateHandler add(final String uriTemplate, final HttpHandler handler) {
+        Objects.requireNonNull(uriTemplate);
+        Objects.requireNonNull(handler);
+
+        // Router builders are not thread-safe, so we need to synchronize.
+        synchronized (lock) {
+            builder.addTemplate(uriTemplate, handler);
+            router = builder.build();
         }
-        match.getValue().handleRequest(exchange);
-    }
 
-    public PathTemplateHandler add(final String uriTemplate, final HttpHandler handler) {
-        pathTemplateMatcher.add(uriTemplate, handler);
         return this;
     }
 
+    /**
+     * Removes a template from the underlying router.
+     *
+     * @param uriTemplate The URI path template.
+     *
+     * @return Reference to this handler.
+     */
     public PathTemplateHandler remove(final String uriTemplate) {
-        pathTemplateMatcher.remove(uriTemplate);
+        Objects.requireNonNull(uriTemplate);
+
+        // Router builders are not thread-safe, so we need to synchronize.
+        synchronized (lock) {
+            builder.removeTemplate(uriTemplate);
+            router = builder.build();
+        }
+
         return this;
     }
 
     @Override
     public String toString() {
-        Set<PathTemplate> paths = pathTemplateMatcher.getPathTemplates();
-        if (paths.size() == 1) {
-            return "path-template( " + paths.toArray()[0] + " )";
+        final List<PathTemplateRouter.PatternEqualsAdapter<PathTemplateRouter.Template<Supplier<HttpHandler>>>> templates;
+        synchronized (lock) {
```

Comment:
toString impeding normal operation is not ideal.

Reply:
Agreed.  I have updated this.  I believe that this toString method is probably only used for debugging purposes due to the size of the string produced (Consistent with the original implementation).  This may now produce results that aren't entirely consistent when called concurrently with routes being mutated, but that is also consistent with how the original implementation worked.